File: Hash/930.Binary-Subarrays-With-Sum/930.Binary-Subarrays-With-Sum.ts
function numSubarraysWithSum(nums: number[], goal: number): number {
  let s = 0;
  let ans = 0;
  const prefix = { 0: 0 };
  for (const i of nums) {
    prefix[s] = (prefix[s] || 0) + 1;
    s += i;
    ans += prefix[s - goal] || 0;
  }
  return ans;
}

File: Hash/930.Binary-Subarrays-With-Sum/930.Binary-Subarrays-With-Sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        Map<Integer, Integer> prefix = new HashMap<>();
        int ret = 0;
        int s = 0;
        for (int i = 0; i < nums.length; i++) {
            prefix.put(s, prefix.getOrDefault(s,0) + 1);
            s += nums[i];
            ret += prefix.getOrDefault(s - goal, 0);
        }
        return ret;
    }
}
